contracts/Registry/TokenCuratedRegistry.sol

```diff
@@ -1,59 +1,97 @@
+  // challenge contract.
+  function challenge(bytes32 id) public {
+    require(exists(id) && !_challengeExists(id));
+    require(token.transferFrom(msg.sender, this, minStake));
+    address challenge = challengeFactory.create(this, msg.sender, owners[id]);
+    require(token.transfer(challenge, minStake.mul(2)));
   }
 
-  constructor (
-    ERC20 _token,
-    uint _minStake,
-    uint _applicationPeriod,
-    ChallengeFactory _challengeFactory
-  ) public
-    StakedRegistry(_token, _minStake)
-  {
-    applicationPeriod = _applicationPeriod;
-    challengeFactory = _challengeFactory;
+  // Handles transfer of reward after a challenge has ended. Requires that there
+  // is an ended challenge for the item.
+  function resolveChallenge(bytes32 id) public {
+    require(_challengeEnded(id));
+    if (_challengePassed(id)) {
+      // if the challenge passed, reward the challenger (via token.transfer) and remove
+      // the item.
+      require(token.transfer(challenges[id].challenger(), _redeemReward(id)));
+      delete ownerStakes[id];
+      super.remove(id);
+    } else {
+      // if the challenge failed, reward the applicant (by adding to their staked balance)
+      ownerStakes[id] = ownerStakes[id].add(_redeemReward(id)).sub(minStake);
+    }
+    delete unlockTimes[id];
+    delete challenges[id];
   }
 
-  function add(bytes32 data) public returns (bytes32) {
-    bytes32 id = keccak256(data);
-    itemsCurationData[id] = ItemCurationData(now + applicationPeriod, false, address(0), false);
-    return super.add(data);
+  // Returns true if the item exists and is not locked. We know that locked items are in
+  // the application phase, because the unlock time is set to now + applicationPeriod when
+  // items are added. Also, unlock time is set to 0 if an item is challenged and the
+  // challenge fails.
+  function inApplicationPhase(bytes32 id) public view returns (bool) {
+    return exists(id) && !isLocked(id);
   }
 
-  function remove(bytes32 id) public itemNotLockedInChallenge(id) {
-    delete itemsCurationData[id];
-    super.remove(id);
+  // internals...
+
+  function _redeemReward(bytes32 id) internal returns (uint reward) {
+    reward = challenges[id].reward();
+    require(token.transferFrom(challenges[id], this, reward));
   }
 
-  function inChallengePhase(bytes32 id) public returns (bool) {
-    ItemCurationData storage itemCurationData = itemsCurationData[id];
+  function _challengeActive(bytes32 id) internal view returns (bool) {
+    return exists(id) && _challengeExists(id) && !challenges[id].ended();
+  }
 
-    if (itemCurationData.challengeAddress == address(0)) {
-      return false;
-    } else if (itemCurationData.challengeResolved) {
-      return false;
-    } else {
-      return true;
-    }
+  function _challengePassed(bytes32 id) internal view returns (bool) {
+    return _challengeEnded(id) && challenges[id].passed();
```

Comment:
I think the outcomes of this function should be as follows:
1. `revert`  Challenge does not exist: none ever existed or all previous challenges are resolved on registry.
2. `revert`  Challenge exists. Challenge has not ended.
3. `true`     Challenge exists. Challenge has ended. Challenge has passed. Challenge has not been resolved on the registry.
4. `false`   Challenge exists. Challenge has ended. Challenge has not passed. Challenge has not been resolved on the registry.

`2.` would check out as long as Challenge contract reverts on `passed()` if it's not ended yet. Wondering  if `1.` checks out. This means either `challenges[id].ended()` or `challenges[id].passed()` needs to revert when called on a `0x0x` address. Do you know if this is the case?